cpp/dolfinx/mesh/graphbuild.cpp

```diff
@@ -276,98 +285,158 @@ graph::AdjacencyList<std::int64_t> compute_nonlocal_dual_graph(
+          matched_facets[facet_b].push_back(cell_a);
+        }
       }
 
-      // Advance iterator and increment entity
-      it = it1;
+      it = matching_facets.end();
     }
   }
 
+  std::vector<int> num_items_po_send;
+  num_items_po_send.reserve(matched_facets.size());
+  std::ranges::for_each(matched_facets, [&](const auto& matches)
+                        { num_items_po_send.push_back(matches.size()); });
+
   // Create neighbourhood communicator for sending data from post
   // offices
   MPI_Comm comm_po_receive;
   MPI_Dist_graph_create_adjacent(comm, dest.size(), dest.data(), MPI_UNWEIGHTED,
                                  src.size(), src.data(), MPI_UNWEIGHTED,
                                  MPI_INFO_NULL, false, &comm_po_receive);
 
-  // Send back data
-  std::vector<std::int64_t> recv_buffer1(send_disp.back());
-  MPI_Neighbor_alltoallv(send_buffer1.data(), num_items_recv.data(),
-                         recv_disp.data(), dolfinx::MPI::mpi_t<std::int64_t>,
-                         recv_buffer1.data(), num_items_per_dest.data(),
-                         send_disp.data(), dolfinx::MPI::mpi_t<std::int64_t>,
-                         comm_po_receive);
-  MPI_Comm_free(&comm_po_receive);
+  // Send po->recipient matched cell counts (non-blocking)
+  std::vector<int> recv_mf_counts(send_disp.back());
+  MPI_Request recv_mf_counts_request;
+  MPI_Ineighbor_alltoallv(num_items_po_send.data(), num_items_recv.data(),
+                          recv_disp.data(), MPI_INT, recv_mf_counts.data(),
+                          num_items_per_dest.data(), send_disp.data(), MPI_INT,
+                          comm_po_receive, &recv_mf_counts_request);
+
+  // Prepare send data for matched facets - unrolled matched arrays.
+  std::vector<std::int64_t> send_mf;
+  std::vector<int> send_mf_sendcounts(num_items_recv.size(), 0);
+  std::vector<std::int32_t> send_mf_displs(send_mf_sendcounts.size() + 1, 0);
+  {
+    send_mf.reserve(
+        std::accumulate(num_items_po_send.begin(), num_items_po_send.end(), 0));
+    for (auto& matches : matched_facets)
+      for (auto match : matches)
+        send_mf.push_back(match);
+
+    int index = 0;
+    for (std::size_t i = 0; i < num_items_recv.size(); i++)
+    {
+      for (int j = 0; j < num_items_recv[i]; j++)
+        send_mf_sendcounts[i] += matched_facets[index + j].size();
 
-  // --- Build new graph
+      index += num_items_recv[i];
+    }
+
+    std::partial_sum(send_mf_sendcounts.begin(), send_mf_sendcounts.end(),
+                     std::next(send_mf_displs.begin()));
+  }
 
-  // Count number of adjacency list edges
-  std::vector<std::int32_t> num_edges(local_dual_graph.num_nodes(), 0);
-  std::adjacent_difference(std::next(local_dual_graph.offsets().begin()),
-                           local_dual_graph.offsets().end(), num_edges.begin());
-  for (std::size_t i = 0; i < recv_buffer1.size(); ++i)
+  // Compute matched facet receive counts and displacements.
```

Comment:
Should the comment be moved to after the MPI_Wait three lines down, as it relates to the for-loop, rather than the assignment of sendcounts?